pandas/core/indexes/multi.py

```diff
@@ -3726,31 +3726,16 @@ def equals(self, other: object) -> bool:
             other_mask = other_codes == -1
             if not np.array_equal(self_mask, other_mask):
                 return False
-            self_codes = self_codes[~self_mask]
-            self_values = self.levels[i]._values.take(self_codes)
-
-            other_codes = other_codes[~other_mask]
-            other_values = other.levels[i]._values.take(other_codes)
-
-            # since we use NaT both datetime64 and timedelta64 we can have a
-            # situation where a level is typed say timedelta64 in self (IOW it
-            # has other values than NaT) but types datetime64 in other (where
-            # its all NaT) but these are equivalent
-            if len(self_values) == 0 and len(other_values) == 0:
-                continue
-
-            if not isinstance(self_values, np.ndarray):
-                # i.e. ExtensionArray
-                if not self_values.equals(other_values):
-                    return False
-            elif not isinstance(other_values, np.ndarray):
-                # i.e. other is ExtensionArray
-                if not other_values.equals(self_values):
-                    return False
-            else:
-                if not array_equivalent(self_values, other_values):
-                    return False
-
+            self_level = self.levels[i]
+            other_level = other.levels[i]
+            new_codes = recode_for_categories(
+                other_codes, other_level, self_level, copy=False
+            )
+            if not np.array_equal(self_codes, new_codes):
+                return False
+            if not self_level[:0].equals(other_level[:0]):
```

Comment:
Could we just use `is_dtype_equal`?

Reply:
The equality rules for numpy dtypes are not as strict as `is_dtype_equal`:

```
In [1]: import pandas as pd

In [2]: import numpy as np

In [3]: idx = pd.Index([1])

In [4]: idx.astype("int64").equals(idx.astype("int32"))
Out[4]: True

In [5]: pd.core.dtypes.common.is_dtype_equal(np.int64, np.int32)
Out[5]: False
```